### PR TITLE
feat(semver): Add function to parse semver search and return a snuba condition representing it (WOR-896)

### PR DESCRIPTION
**Pull request by wedamija**
This implements a function that understands basic semver comparison and sorting, and uses it to
return a snuba condition to fetch events/transactions from snuba using a list of release versions.

This isn't plugged into the search yet, I will do that in a separate pr. It's likely that parts of
this function will need to be extracted and moved to common places so that they can be shared with
`Release` search etc.

We also still need to handle wildcard matching etc, which will come later.

The idea with this function is that we want to return all versions associated with a given semver
condition. For orgs with many `Releases` this can prove to be a lot of rows, so we return a max of
1000 releases. If we hit this max, we also try inverting the query and see if that returns a smaller
number of rows. If it does, we use those rows instead as a `NOT IN` query. Otherwise, we return the
1000 releases we fetched, which are ordered so that they're as close to the passed filter as
possible, which hopefully gives the user the results they're looking for regardless.

### CODE REVIEW

**Review by wmak — APPROVED** [review]
this LGTM, my biggest concern is around the initial 1000 for `MAX_SEMVER_SEARCH_RELEASES` being too large, but from our convo in slack the majority of semver releases aren't long enough for this to be a concern

File: src/sentry/search/utils.py [hunk 1]
```diff
@@ -277,7 +277,9 @@ def get_latest_release(projects, environments, organization_id=None):
     )
 
 
-def parse_release(value, projects, environments, organization_id=None):
+def parse_release(
```
Comment (ahmedetefy) [inline src/sentry/search/utils.py:280]: Nit: I would split this out as part of a different PR since it is not really related

File: src/sentry/search/events/filter.py [hunk 2]
```diff
@@ -321,6 +325,118 @@ def _team_key_transaction_filter_converter(
...
+
+def _flip_field_sort(field: str):
+    return field[1:] if field.startswith("-") else f"-{field}"
+
+
+def parse_semver_search(
+    search_filter: SearchFilter,
+    name: str,
+    params: Optional[Mapping[str, Union[int, str, datetime]]],
+) -> Tuple[str, str, Sequence[str]]:
```
Comment (ahmedetefy) [inline src/sentry/search/events/filter.py:340]: Nit: This function is kinda verbose, maybe not a bad idea to break it down ? like the inverse query could be in its own func and also feel like the part where `major, minor, ... etc` are extracted and then `filter_func` is created could be extracted into its own func too since `major, minor,... etc` are not used beyond https://github.com/getsentry/sentry/pull/26440/files#diff-a278bb701c83c69212137a1d12ead41fa91d73b5305dc4b528233db542811b69R380

**Review by ahmedetefy — APPROVED** [review]
LGTM

File: src/sentry/search/events/filter.py (hunk 2 above)
Comment (wedamija) [reply src/sentry/search/events/filter.py:340]: Yeah, I plan to break it into smaller pieces as I use it in other places. Release search will need to use parts of this without the conversion to a snuba query. I'd rather tackle this as I work on those pieces though, so that I understand the requirements better

File: src/sentry/search/utils.py (hunk 1 above)
Comment (wedamija) [reply src/sentry/search/utils.py:280]: Tbh I'm not sure why I did this at all, wondering if my IDE did it, since marking these all as `object` isn't helpful
